lld/ELF/InputFiles.cpp

```diff
@@ -1418,6 +1439,38 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
   return verneeds;
 }
 
+// To determine if a shared file can support any of the GNU Attributes,
+// the .note.gnu.properties section need to be read. The appropriate
+// location in memory is located then the GnuPropertyNote can be parsed.
+// This is the same process as is used for readGnuProperty, however we
+// do not pass the data variable as, without an InputSection, its value
+// is unknown in a SharedFile. This is ok as the information that would
+// be collected from this is irrelevant for a dynamic object.
+template <typename ELFT>
+void SharedFile::parseGnuAndFeatures(const uint8_t *base,
+                                     const typename ELFT::PhdrRange headers,
+                                     const typename ELFT::Shdr *sHeader) {
+  if (numElfPhdrs == 0 || sHeader == nullptr)
+    return;
+  uint32_t featureAndType = ctx.arg.emachine == EM_AARCH64
+                                ? GNU_PROPERTY_AARCH64_FEATURE_1_AND
+                                : GNU_PROPERTY_X86_FEATURE_1_AND;
+
+  for (unsigned i = 0; i < numElfPhdrs; i++) {
+    if (headers[i].p_type != PT_GNU_PROPERTY)
+      continue;
+    const typename ELFT::Note note(
+        *reinterpret_cast<const typename ELFT::Nhdr *>(base +
+                                                       headers[i].p_vaddr));
```

Comment:
Can you check if p_vaddr is the right field? I think it should be p_offset? For most shared-libraries these will be the same for the PT_GNU_PROPERTY, but it is not guaranteed.

The p_vaddr is the virtual address, the p_offset is the file offset within the ELF file. The p_vaddr is meant to be used after the file has been memory mapped into virtual memory, when operating on the file p_offset should be used.

Most shared libraries start at address 0x0 but this is not a requirement so I'd expect using `p_vaddr` to break if a different base address had been used.

See https://www.sco.com/developers/gabi/latest/ch5.pheader.html